### DISCUSSION TITLE
[CHAPTER 4] QnA

### DISCUSSION THREAD

**Discussion by BvrPark** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#42] [CHAPTER 4 스트림 소개 - Q&A]
### 질문 1.
스트림과 컬렉션의 차이점에서 컬렉션은 적극적 생성, 스트림은 게으른 생성을 한다고 합니다.
그래서 컬렉션은 현재 자료구조에 포함된 모든 값을 계산한 다음 컬렉션에 추가할 수 있다고 하는데 이게 어떤 뜻일까요?
그리고 여기서 현재 자료구조라는 것이 무엇을 뜻할까요? 단순히 우리가 아는 배열, 스택, 큐 이런 자료구조는 아닌 것 같아서요.

**Comment by ahngilwoong** [comment]
스트림의 장점 중 하나가 스트림에서 중간 연산이 수행될 때는 요소들이 실제로 처리되지 않고 미뤄놨다가 최종 연산시에만 필요한 데이터만 처리하고 불필요한 데이터는 연산하지 않는데 실제로 연산을 바로바로 처리하지 않는 점이 게으른 방식을 쓴다고 말할 수 있는 것 같습니다. 필요한 데이터만 처리하면 자원의 낭비가 줄어들게 됩니다! 

자료구조는 컴퓨터 과학에서 효율적인 접근 및 수정을 가능케 하는 자료의 조직, 관리, 저장을 의미한다.  더 정확히 말해, 자료 구조는 데이터 값의 모임, 또 데이터 간의 관계, 그리고 데이터에 적용할 수 있는 함수나 명령을 의미한다. 라고 적혀있습니다.
컬렉션도 데이터의 저장, 조작, 관리등을 하므로 데이터를 관리하기 위한 자료구조의 한 형태로 볼 수 있다고 생각합니다. 그래서 말씀하신 우리가 아는 그런 부분들과 동일한 말을 뜻하는 듯 합니다!

**Reply by BvrPark** [reply]
저도 처음에 컬렉션도 자료구조의 한 형태라고 생각하고 읽어봤는데
그러면 `컬렉션은 현재 자료구조에 포함된 모든 값을 계산한 다음 컬렉션에 추가할 수 있다.`
라는 말이 이해가 잘 가질 않습니다.

컬렉션을 자료구조의 한 형태로 생각하면
`컬렉션은 컬렉션에 포함된 모든 값을 계산한 다음 컬렉션에 추가할 수 있다`
이렇게 해석될 수 있는데 그러면 컬렉션에 포함된 값을 다 계산해야지 컬렉션에 추가할 수 있다?
컬렉션에 추가를 해야 계산도 할 수 있는 것이 아닌지 궁금합니다!

**Reply by ahngilwoong** [reply]
아 말씀하신 것이 제가 이해한게 맞다면 느낀대로만 얘기해보겠습니다.
예를 들어서 1부터 100까지 짝수의 합을 계산한다고 치면 

1. 컬렉션의 경우
1부터 100까지의 모든 값을 메모리에 적재 후 짝수 값만 골라서 더한다.

2. 스트림의 경우
실제 필요한 값(짝수 값)만 찾아서 더한다.

로 저는 이해했습니다.
그래서 제가 이해한게 맞다면 책에서 계산한다는 말은 메모리에 적재되는 과정을 계산하여 메모리에 적재한다 라고 말한다고 생각하고
메모리에 적재 후 추가를 한다? 라고 느껴집니다.
